### ISSUE TITLE
prebrowsing javascript api

### ISSUE THREAD

**Issue by Jxck** [w3c/resource-hints#36]
like single page application, there cause lots of dynamic fetch request from javascript.
but if we need to optimize them using pre-*.

we need to scan all javascript source and apply to html as link tag,
or create link element from javascript and add it to dom.

I think it's helpful to export javascript API for pre-*
we don't need to care about sync with dom link tag,
and complete in javascript source.

for example
in case of send a request with search query and fetch result.
we can pre-connect to search server while user inputs query.

like rails turbo-links style pjax.
prefetch, prerender will fit.

I think someone says service worker + fetch + cache will help.
but we can't do preconnect, we don't wanna send http request.
we also can't do prerender, service worker doesn't render it.
prefetch seems work, but we need to write heavy code and register to browser.
easy way for enable them to browser API are welcome.

**Comment by toddreifsteck** [comment]
A link can be added using javascript that contains a preconnect or other resource hint tag. Is that a sufficient to cover this request?

**Comment by igrigorik** [comment]
@Jxck check out Example 7: http://w3c.github.io/resource-hints/#fetching-the-resource-hint-link

**Comment by Jxck** [comment]
@toddreifsteck @igrigorik why we need DOM for optimize browser networking ? (as i already said at first issue comment, please check out)

**Comment by igrigorik** [comment]
@Jxck I'm not 100% on what you're asking, but.. You _can_ script preconnect, preload, prefetch, and prerender via JavaScript via the mechanism in example 7; for bonus points, you can wrap that code into a "nicer" API that best suits your application. Also, where you get the list of resources for which you want to trigger these operations is completely open: the server can write them out as HTML markup or Link headers, which will help optimize the initial load of the page; you can keep the references in JS-land and trigger those requests whenever your code deems it necessary. 

But, perhaps to answer your last question more directly: we need declarative HTML markup because it allows us to invoke these optimizations on initial load of the page, without being blocked on JavaScript - i.e. the preload scanner can execute them even if/when the doc parser is blocked on blocking JS.

**Comment by toddreifsteck** [comment]
I believe that @Jxck is asking for a javascript API that does not touch the DOM.

The fetch API should eventually provide this for many of the resource hints, but preconnect is underneath fetch.

Imagine an API such as:

a.       preconnect(“//foo.comhttp://foo.com”)

Admittedly, this could be exposed as a polyfill to <link rel=”preconnect” href=”//foo.com”> but it may be purer to expose this on ServiceWorker or just in Window or WorkerGlobalScope.

**Comment by igrigorik** [comment]
Ah, I see, that makes sense. I'll defer to @annevk on how (and if) to go about that.. since this seems like yet another property on fetch().

**Comment by Jxck** [comment]
thanks @igrigorik for answer, and thanks @toddreifsteck for following me.

> you can wrap that code into a "nicer" API that best suits your application. 

yes, if doing that like Example 7, we can do that eventually.
but that's the same thing to say `we don't need XHR, you can create <a> and click it and get result from DOM, so wrap it.` isn't it ? (really but example lol)

and also, in (web|shared|service) worker, we can't touch to DOM.
that means we don't have chance to optimize `fetch()` operation in way of Example 7

> we need declarative HTML markup 

that's a important point so **I don't against to Link tag** itself.
in extensible way, if we have a same meaning api in javascript,
we can polyfill Link tag behavior, BUT pre-\* is sensitive to execute timing.
so, I need Link tag way, for example that's works optimize fetching subresources 
from static content. 

but we also have many chance to optimize fetching from javascript,
in some case, server doesn't know about hostname or URL.
but doing same thing from javascript like Example 7 seems half measure for me.
because I really wanna do is
- `preconnect` : make a connection from browser to server and use it next fetching from browser
- `prefetch`: fetch a resource and cache to browser and user it next fetching from browser
- `prerender`: fetch and render and (same thing here)

all we wanna do is not **create a Link element** but do same network optimize.

> Imagine an API such as: `preconnect(“//foo.com<http://foo.com>”)`

yes, that's a exactly kind of I wanna say!! :)

> but it may be purer to expose this on ServiceWorker or just in Window or WorkerGlobalScope.

fetching from `fetch()` or XHR or normal location change is quite natural for recent application, so force registering service worker for optimizing `pre-*` is seems quite heavy for me.
and also, if we need service worker for doing `pre-*` things, we can't do that until service worker registerd. that's reduces chance and especially we can't optimize service worker download itself. 

> since this seems like yet another property on fetch().

that's seems reasonable, but `fetch()` itself is a single function and only for fetch().
so in that case, we need new other API or so.

**Comment by annevk** [comment]
I think the reason we don't have an API here is because you want the HTML's parser's scanner to find these links and that would not work if you have to execute script first.

**Comment by igrigorik** [comment]
@annevk right. @Jxck is pointing out that you can't issue a preconnect from a Worker given current API & processing model - e.g. I'd like to preconnect to an origin from my ServiceWorker, but within SW I don't have access to the DOM, so.. I can't insert a <link rel>, etc.

**Comment by annevk** [comment]
Right, but service workers are short-lived... User agents analyzing what service workers typically connect to and warming those up might work. Perhaps providing some kind of `preload` dictionary when you install the service worker that you can update in some way? So that while the user agent starts running the service worker it can also start warming up connections.

There've been a number of ideas here, but I'm not sure a new API is it.

**Comment by igrigorik** [comment]
@annevk why is "short-lived" a concern? In the case of a preconnect, the socket lifecycle is completely independent of any particular ServiceWorker, and for prefetch/preload the response can still go into the regular HTTP cache. 

Concretely, imagine I want to build a [site-optimized preconnect predictor](https://www.igvita.com/posa/high-performance-networking-in-google-chrome/#predictor): to do so, I want to observe fetches (via SW) as they fly by and initiate preconnects for apprioriate origins. For example, anytime I see a request to 'https://fonts.googleapis.com', I want to fire a preconnect for 'https://fonts.gstatic.com', etc. Similar logic applies to preload, prefetch, etc.

To enable this, we would need a couple of new JS API's that need to be accessible from SW...
- rel=preconnect --> navigator.preconnect(origin), or some such.
- rel=preload --> noop, this is simply fetch()
- rel=prefetch --> fetch with low priority (gets back to being able to communicate priority to fetch)
- rel=prerender --> perhaps this one needs a separate API as well, since it spawns a new render process and all that. navigator.prerender(url), maybe?

`</crazytalk>`

**Comment by Jxck** [comment]
@annevk, @igrigorik

> you want the HTML's parser's scanner to find these links

No, target URL doesn't always in HTML.
some application creates URL dynamically from user interaction.
If url for pre-\* are already wrote in HTML at first,
we can wrote Link tag with that too, and no JS API required.

> Perhaps providing some kind of preload dictionary when you install the service worker that you can update in some way?
> 
> To enable this, we would need a couple of new JS API's that need to be accessible from SW...

I think the new API for pre-\* should not premise service worker,
this API should accessible from `window` and `self`(sw), like `fetch()`

for example, sw hooks fetching as fetch event, but it's a post processing of fetching happend.
but in pre-\* cases, we need to do them as pre processing (as their name), so there is no event kicks sw. there is no reason to do it only from service worker.

and more, we can't optimize networks before service worker registering. we can't optimise service worker download itself.

> rel=preconnect --> navigator.preconnect(origin), or some such.

lgtm.

> rel=preload --> noop, this is simply fetch()
> rel=prefetch --> fetch with low priority (gets back to being able to communicate priority to fetch)

response of `fetch()` is cached in browser and next clicking A tag will hit the cache ?
or you saying about using Cache API ?

> rel=prerender --> perhaps this one needs a separate API as well, since it spawns a new render process and all that. navigator.prerender(url), maybe?

maybe.

**Comment by igrigorik** [comment]
I'd be curious to hear @slightlyoff's thoughts on how to "explain `prerender`" in the context of SW and fetch. :)

**Comment by Jxck** [comment]
I wanna hear that too.

**Comment by yoavweiss** [comment]
Given the inactivity for the last 6+ years, closing. Feel free to re-open if there's still interest in this.
